0000-experiment-dataframe.md

```diff
@@ -0,0 +1,320 @@
+
+```python
+sorted_frame = result_dataframe.sort_values(by=["creation_datetime"])
+```
+
+alternatively one can directly visualize the data frame with seaborn:
+
+```python
+import seaborn as sns
+
+sns.relplot(
+    data=result_dataframe,
+    kind="line",
+    x="creation_datetime",
+    y="value",
+)
+```
+
+See pandas documentation for more details.
+
+### B. Introduction of artifacts
+
+Result database allows experimentalists to store `artifact` data per experiment ID. Thus, in QE, we can add `ExperimentData.artifact` to save supplementary data such as raw curve data points. In principle this can be a free-from dictionary, but it would be better to provide the artifact base class and subclasses to support custom JSON serialization to generate the REST API payload and the input data type validation. In the following we describe `CurveAnalysisArtifact` for `CurveAnalysis`. Note that artifact must be the [thread-safe container](https://github.com/Qiskit/qiskit-experiments/blob/b150427f86c74a02d638e06507864c7bb060c27c/qiskit_experiments/database_service/utils.py#L168-L245) subclass. Each analysis class must know the associated artifact subclass to populate the supplementary data.
+
+The artifact is a part of `ExpeirmentData`, however, each `BaseAnalysis._run_analysis` method must instantiate and return a list of artifact instances to the experiment data. For example, a composite analysis instance may return multiple artifacts for each qubit or experiment.
+
+
+#### BaseArtifact
+
+The base class of the artifact. Probably Python `dataclass`. This class implements `__json_encode__` and `__json_decode__` methods and subclass can override if necessary.
+This base class may have the field for list of figures. In addition to this, artifact object must provide metadata including
+- experiment (str): Name of experiment that generated this data.
+- device_components (List[DeviceComponent]): List of device components that this data is associated with.
+- creation_datetime (datetime): When this data is created. For example, an experimentalists can sort the artifacts by date to visualize the trend of curves in a heat map in line with time.
+
+
+#### CurveAnalysisArtifact
+
+This artifact have two extra fields.
+
+- fit_status: A dictionary (that may provide nice html view) of fitting status. This is currently provided as analysis result entry with the name `@Parameters_*`. This is not really the analysis results and thus moved to artifact. This contains information about fitting, such as fit models, solver status, initial and final fit parameters and so forth. For example,
```

Comment:
Im not confident subclasses is the right way to go for artifacts, it looks more like here you want to define some class that would itself be the value/data of an artifact that is saved, and you want some nice way to html visualize that value.

Reply:
Probably we have assume different data structure for artifact. I assumed following:
```
Artifact: List[BaseArtifact]
```
and 
```
BaseArtifact: dataclass
```
where single analysis instance returns a single formatted artifact object. So data type validation is done by each artifact instance and artifact can contain multiple data (e.g. figure, curve data, fit status, ...).

Likely you are assuming following:
```
Artifact: List[Any]
```
where a single analysis instance generates multiple data entries, and it can push them to the artifact without structuring the data. Then each entry is managed by the experiment id and name. For example, parallel experiment of the same kind of experiment can push merged curve data as a single data entry. The artifact object does validation for acceptable data type. I like your approach.

Reply:
I redesigned the artifact. Seems like new approach is better than that I originally proposed here. https://github.com/Qiskit/rfcs/pull/28/commits/263e94a3c5a02e3be14c7916652590b63e970ba7